pkg/webhooks/podimagespec/podimagespec.go

```diff
@@ -0,0 +1,281 @@
+		return true, nil
+	} else if s.configV1.Spec.ManagementState == operatorv1.Removed {
+		return false, nil
+	}
+
+	return false, fmt.Errorf("unknown managementState: %s", s.configV1.Spec.ManagementState)
+}
+
+// Authorized implements Webhook interface
+func (s *PodImageSpecWebhook) Authorized(request admissionctl.Request) admissionctl.Response {
+	//Implement authorized next
+	return s.authorizeOrMutate(request)
+}
+
+// authorizeOrMutate decides whether the Request requires mutation before it's allowed to proceed.
+func (s *PodImageSpecWebhook) authorizeOrMutate(request admissionctl.Request) admissionctl.Response {
+	var ret admissionctl.Response
+	pod, err := s.renderPod(request)
+	if err != nil {
+		log.Error(err, "Couldn't render a Pod from the incoming request")
+		return admissionctl.Errored(http.StatusBadRequest, err)
+	}
+	//1. Regex match
+	for i := range pod.Spec.Containers {
+		container := &pod.Spec.Containers[i]
+		matched, namespace, _, _, err := checkContainerImageSpecByRegex(container.Image)
+		if err != nil {
+			return admissionctl.Errored(http.StatusBadRequest, err)
+		}
+		if namespace != "openshift" {
+			ret.UID = request.AdmissionRequest.UID
+			ret = admissionctl.Allowed("Pod image spec is valid")
+			return ret
+		}
+		//if the regex matches, check image-registry status
+		if matched {
+			ctx := context.Background()
+			//check if image-registry is enabled
+			registryAvailable, err := s.CheckImageRegistryStatus(ctx)
+			if err != nil {
+				log.Error(err, "Failed to check image registry status")
+				return admissionctl.Errored(http.StatusInternalServerError, err)
+			}
+			if !registryAvailable {
+				imageMutated, err := s.mutateContainerImageSpec(&container.Image, namespace, ctx)
+				if err != nil {
+					return admissionctl.Errored(http.StatusBadRequest, err)
+				}
+				if imageMutated {
+					ret = admissionctl.Patched(
+						fmt.Sprintf("images on pod %s mutated for reliability", pod.GetName()),
+					)
+
+					log.Info(fmt.Sprintf("images on pod %s mutated for reliablity", pod.GetName()))
+					// ret.Complete() sets the UID and finalizes the patch
+					ret.Complete(request)
+				} else {
+					return admissionctl.Errored(http.StatusBadRequest, err)
+				}
+			} else {
+				ret = admissionctl.Allowed("Image registry is available, no mutation required")
+				ret.UID = request.AdmissionRequest.UID
+			}
+		} else {
+			ret = admissionctl.Allowed("Pod image spec is valid")
+			ret.UID = request.AdmissionRequest.UID
+		}
+	}
+
+	return ret
+}
+
+// renderPod renders the Pod in the admission Request
+func (s *PodImageSpecWebhook) renderPod(request admissionctl.Request) (*corev1.Pod, error) {
+	decoder, err := admissionctl.NewDecoder(s.s)
+	if err != nil {
+		return nil, err
+	}
+	pod := &corev1.Pod{}
+	if len(request.OldObject.Raw) > 0 {
+		err = decoder.DecodeRaw(request.OldObject, pod)
+	} else {
+		err = decoder.DecodeRaw(request.Object, pod)
+	}
```

Comment:
```suggestion
	err = decoder.Decode(req, pod)
```
I made the mistake of doing this `OldObject` checking in my original version of this webhook and have since patched it. See #298 for details, but tl;dr is don't bother with `OldObject` unless you want really difficult-to-reproduce bugs. You might also want to see the ways I updated the unit tests in that PR